version/cmd/version/version.go

```diff
@@ -56,34 +40,54 @@ func main() {
 		flag.PrintDefaults()
 		os.Exit(1)
 	}
-	result, err := fn(tag)
+	result, err := fn(tag, bt)
 	if err != nil {
 		log.Fatal(err)
 	}
 	fmt.Println(result)
 }
 
-func getTag() (string, error) {
+// buildType getter func that matches the signature of other version funcs
+func buildType(_ string, bt version.BuildType) (string, error) {
+	return string(bt), nil
+}
+
+// implements the version.BuildEnv interface using the real build environment
+type BuildEnv struct{}
+
+// Returns true if we are building from a CI (rather than a local or dev build)
+func (b *BuildEnv) IsCI() bool {
+	// Travis, AppVeyor, CircleCI, and many others all set CI=true in their
+	// environment by default.
+	return os.Getenv("CI") == "true"
+}
+
+// Returns true if this is a nightly release by checking whether the current
+// HEAD is one or more commits ahead of the latest tag.
+func (b *BuildEnv) IsNightly() bool {
 	cmd := exec.Command("git", "describe", "--exact-match", "--tags", "HEAD")
-	out, err := cmd.Output()
-	if err != nil {
-		if _, ok := err.(*exec.ExitError); !ok {
-			return "", err
-		}
+	err := cmd.Run()
+	// if the tag is an exact match for current HEAD, this is not a nightly
+	if err == nil {
+		return false
 	}
-	tag := strings.Trim(string(out), "\n")
-	return tag, nil
+	// if the command exited with a nonzero status, this is a nightly build
+	if _, ok := err.(*exec.ExitError); ok {
+		return true
+	}
+	// if the command somehow failed to execute, it is a fatal error
+	log.Fatal(err)
+	return false
 }
 
-func getTags() ([]string, error) {
-	cmd := exec.Command("git", "tag", "-l")
+// Returns the most recent tag belonging to the current commit
+func (b *BuildEnv) GetMostRecentTag() string {
+	// --abbrev=0 disables tag annotation, returning unmodified most recent tag
+	cmd := exec.Command("git", "describe", "--abbrev=0", "--tags", "HEAD")
 	out, err := cmd.Output()
 	if err != nil {
-		if _, ok := err.(*exec.ExitError); !ok {
-			return nil, err
-		}
+		log.Fatal(err)
```

Comment:
`log.Fatal` is generally best used only in main. The function is a log.Println followed by os.Exit.

That means deferred functions will never be run, and it may be hard to determine the source of the error.

The error should be part of the method and passed through.

Reply:
These functions define a struct that satisfies the `BuildEnv` interface for use by main, and only by main, which is defined in the same file. Therefore I've considered them to be conceptually part of main (from an abstraction layer perspective) even if technically they are separate.

I was not aware of the implications of using `log.Fatal` exiting without any possibility of cleanup though. I'll update this propagate errors to callers instead.